pybamm/models/submodels/particle/polynomial_many_particles.py

```diff
@@ -0,0 +1,299 @@
+        ]
+        j = variables[self.domain + " electrode interfacial current density"]
+        T = variables[self.domain + " electrode temperature"]
+        R = variables[self.domain + " particle distribution in x"]
+        if self.order == 0:
+            # No algebraic equations since we only solve for the average concentration
+            pass
+        elif self.order == 2:
+            # We solve an algebraic equation for the surface concentration
+            if self.domain == "Negative":
+                self.algebraic = {
+                    c_s_surf: self.param.D_n(c_s_surf, T) * (c_s_surf - c_s_rav)
+                    + self.param.C_n * (j * R / self.param.a_n / 5)
+                }
+
+            elif self.domain == "Positive":
+                self.algebraic = {
+                    c_s_surf: self.param.D_p(c_s_surf, T) * (c_s_surf - c_s_rav)
+                    + self.param.C_p * (j * R / self.param.a_p / self.param.gamma_p / 5)
+                }
+        elif self.order == 4:
+            # We solve a different algebraic equation for the surface concentration
+            # that accounts for the average concentration gradient inside the particle
+            q_s_rav = variables[
+                "R-averaged " + self.domain.lower() + " particle concentration gradient"
+            ]
+            if self.domain == "Negative":
+                self.algebraic = {
+                    c_s_surf: self.param.D_n(c_s_surf, T)
+                    * (35 * (c_s_surf - c_s_rav) - 8 * q_s_rav)
+                    + self.param.C_n * (j * R / self.param.a_n)
+                }
+
+            elif self.domain == "Positive":
+                self.algebraic = {
+                    c_s_surf: self.param.D_p(c_s_surf, T)
+                    * (35 * (c_s_surf - c_s_rav) - 8 * q_s_rav)
+                    + self.param.C_p * (j * R / self.param.a_p / self.param.gamma_p)
+                }
+
+    def set_initial_conditions(self, variables):
+        c_s_rav = variables[
+            "R-averaged " + self.domain.lower() + " particle concentration"
+        ]
+
+        if self.domain == "Negative":
+            x_n = pybamm.standard_spatial_vars.x_n
+            c_init = self.param.c_n_init(x_n)
+
+        elif self.domain == "Positive":
+            x_p = pybamm.standard_spatial_vars.x_p
+            c_init = self.param.c_p_init(x_p)
+
+        self.initial_conditions = {c_s_rav: c_init}
+
+        if self.order in [2, 4]:
+            # We also need to provide an initial condition (initial guess for the
+            # algebraic solver) for the surface concentration
+            c_s_surf = variables[self.domain + " particle surface concentration"]
+            self.initial_conditions.update({c_s_surf: c_init})
+        if self.order == 4:
+            # We also need to provide an initial condition (initial guess for the
+            # algebraic solver) for the average concentration gradient
```

Comment:
I will remove the comment about algebraic solver here, as it is an ODE for q (accidentally left the comment here from copy/paste above)